feat: support batch delete snapshots

Tql-ws1 opened the pull request:
#5

hirak99 commented on src/code/main.py:
```diff
@@ -107,7 +134,46 @@ def _delete_snap(configs_iter: Iterable[configs.Config], path_suffix: str, sync:
+    indicator: str,
+    sync: bool,
+):
+    configs_iter = list(configs_iter)
```
nit: Change the argument to configs_iter: list[configs.config]?
Or alternatively, instead of re-assigning in place, create a new `configs_list = list(configs_iter)` here.
Reason - mostly for code legibility purposes, makes it clear that the new variable is a list. Theoretical drawback that python will keep the old config_iter is pretty negligible. You can choose to `del configs_iter` after so that it is not inadvertently used afterwards.

Tql-ws1 replied:
Thank you for pointing that out. The change of the data type of `configs_iter` from `Iterable[configs.config]` to `list[configs.config]` is indeed confusing. No problem, I will change `configs_iter = list(configs_iter)` to `configs_list = list(configs_iter)`.

hirak99 commented on src/code/main.py:
```diff
@@ -17,6 +17,7 @@
 import datetime
 import logging
 from typing import Iterable
+from pathlib import Path
```
nit: Following the Google python style guide, import pathlib and use pathlib.Path.

(typing imports are an exception to that rule.)

Tql-ws1 replied:
Got it.

Tql-ws1 commented on src/code/snap_operator.py:
```diff
@@ -54,6 +56,170 @@ def find_target(config: configs.Config, suffix: str) -> Optional[snap_holder.Sna
+    return mapping
+
+
+def _convert_to_timestamp(suffix: str) -> str:
+    """Convert human-readable time format into a timestamp."""
+    if not _human_friendly_timestamp(suffix):
+        return suffix
+
+    try:
+        # "2024-11-19_15:30".split("_") -> ["2024-11-19", "15:30"]
+        date, time = suffix.split("_")
+    except ValueError:  # unpack error
+        date = suffix
+        # If the user does not provide the hour and minute, it defaults to 23:59
+        hour_minute = "2359"
+    else:
+        # "15:30".split(":") -> ["15", "30"] -> "1530"
+        # "1530".split(":") -> ["1530"] -> "1530"
+        hour_minute = "".join(time.split(":"))
+        suffix_not_minutes = len(hour_minute) == 2
+        if suffix_not_minutes:
+            hour_minute += "00"
+
+    days_not_padded_zeros = r"^\d{4}-\d{2}-\d$"
+    pad_days_with_zeros = r"^\d{4}-\d{2}-0\d$"
+    year_month_day = re.sub(days_not_padded_zeros, pad_days_with_zeros, date)
+    return year_month_day + hour_minute
```
## Premise

Assume there is 1 configuration and 5 snapshots:

```txt
Config: xxx/y.conf
Snaps:
  1 202412131055    U
  2 202412132359  S

  3 202412140810    U
  4 202412142359  S

  5 202412151756    U
```

Using `--start TIMESTAMP`, `--end TIMESTAMP`, and `--start TIMESTAMP1 --end TIMESTAMP2`:

1. It should delete all snapshots, but actually deletes snapshot number `2` and all subsequent snapshots.
    - `yabsnap batch-delete --start 2024-12-13` (_23:59)\
    The ideal timestamp completion should be:
    - `yabsnap batch-delete --start 2024-12-13` (_00:00)
2. It should delete snapshots before number `4`, and it actually deletes snapshots before number `4` (the timestamp completion here is correct).
    - `yabsnap batch-delete --end 2024-12-14` (_23:59)
3. It should delete snapshots `1-3`, but actually deletes `2` and `3`.
    - `yabsnap batch-delete --start 2024-12-13 --end 2024-12-14`\
    This is equivalent to:
    - `yabsnap batch-delete --start 2024-12-13_23:59 --end 2024-12-14_23:59`\
    The ideal completion for the above shorthand dates should be:
    - `yabsnap batch-delete --start 2024-12-13_00:00 --end 2024-12-14_23:59`

## Issue

After reviewing the above examples, I believe you can understand that the current `_convert_to_timestamp` function has flaws.

It has the following issues:

- When the timestamp provided by the user has a precision of at most days (e.g. `2024-01-01`), the timestamp completion is inaccurate.
- The design for zero-padding of months and days, as well as the completion of minutes, is not well-implemented (the following are simplified examples):
  - `2024-5-01` -> `2024-05-01`
  - `2024-01-5` -> `2024-01-05`
  - `2024-01-01_02` -> `2024-01-01_02:00`

## Thoughts

When writing the `_convert_to_timestamp` function, I did not consider the introduction of state. When the timestamp provided by the user has a precision of at most days, it is necessary to know which command-line optional parameters (`--start`, `--end`) the user has used so that the `_convert_to_timestamp` function can provide accurate timestamp completion.

Currently, regardless of which optional parameter (`--start`, `--end`) is used, the timestamp completion is always `_23:59`, which is unreasonable. However, introducing state would increase the complexity of the code, which exceeds my expectations.

My thought is to abandon the implementation of "**diverse** human-friendly timestamp parameters" and instead limit users to only input specific formats of timestamps, raising the upper limit to reduce the complexity of functionality implementation and avoid unintentional invalid input from users. What are your thoughts and suggestions on this?

Tql-ws1 replied:
Then I will follow my own idea to implement it: the human-friendly timestamp format will only support the ISO 8601 standard.

Tql-ws1 replied:
> 100% agree. We can choose to make it more complicated later if we need to.

😂

> As for the format we demand from the users, WDYT of using the same format as is shown in "yabsnap list", i.e. YYYYMMDDhhmmss, e.g. --start=202412131055 --end=202412142359? It will keep it consistent, while avoiding confusion.

In some cases, it's difficult for me to visually separate the numbers that are stuck together and associate them with dates, times, etc. So, I still kept the option "allow input in ISO 8601 date format."

> Another thought - if not now then later, I feel we should have a _is_within_timescope(to_checl:datetime, start: str, end: str). This is the logical unit that find_multi_targets() can directly rely on. Also this is the intricate part, and this is the function that can/should be unit-tested (e.g. with similar examples as you have stated in your comment).

I think [TimeScopeFilter](https://github.com/Tql-ws1/yabsnap/commit/f0eb2a69faa94900637ccaf1a5a8465b21f020a4#diff-aca5df2f16d84600a950ca86b8f64bdf8c61ef6ff0809ba813736c1a7af61bd7R58) has already achieved this 👀

hirak99 reviewed (approved):
I think the main steps remaining are to move the relevant functions to a new module (e.g. `batch_deleter.py`), and to design and document the time format by the user.

If you prefer, we can merge this in and I can also help with some of those.

Thank you for your contribution to add this feature.

Tql-ws1 commented:
> I think the main steps remaining are to move the relevant functions to a new module (e.g. `batch_deleter.py`), and to design and document the time format by the user.

I have moved the code for the new feature into `batch_deleter.py`, except for `snap_operator.config_snapshots_mapping` (previously named `_snapshots_of_configs`). 

`config_snapshots_mapping` is a key component of the new functionality and relies on `snap_operator._get_old_backups`. However, because `_get_old_backups` is private, I am hesitant to move `config_snapshots_mapping` to `batch_deleter.py`. Could you please instruct me whether it's possible to make `_get_old_backups` a public function? If not, I will have to duplicate the code in `batch_deleter.py`, which would be redundant.

Additionally, I wrote unit tests related to the new feature. However, there is now a significant issue:
After moving the new feature's code into `batch_deleter.py`, I also modified the data model returned by `snap_operator.config_snapshots_mapping` (previously named `_snapshots_of_configs`). The data structure changed from `dict[str, dict[datetime.datetime, snap_holder.Snapshot]]` to `dict[configs.Config, list[snap_holder.Snapshot]]`.

This change simplifies the data model and makes ideal data operations easier. However, I did not realize that `configs.Config` is unhashable, rendering all operations based on this new data structure invalid.

Now, I need to rethink how to address these issues.
I have committed the relevant code [here](https://github.com/Tql-ws1/yabsnap/commit/f0eb2a69faa94900637ccaf1a5a8465b21f020a4).

Tql-ws1 commented on src/code/snap_operator.py:
```diff
@@ -54,6 +56,170 @@ def find_target(config: configs.Config, suffix: str) -> Optional[snap_holder.Sna
+        # Delete snapshots within a certain time range (inclusive of the left endpoint and exclusive of the right endpoint).
+        case (start, end) if start and end:
+            start_datetime = datetime.datetime.strptime(start, global_flags.TIME_FORMAT)
+            end_datetime = datetime.datetime.strptime(end, global_flags.TIME_FORMAT)
```
Uh... I thought `match ... case...` would have scoping issues. I learned something new!

Tql-ws1 commented on src/code/snap_operator.py:
```diff
@@ -54,6 +56,170 @@ def find_target(config: configs.Config, suffix: str) -> Optional[snap_holder.Sna
+    """List the snapshots to be deleted and ask if they really want to delete them."""
+    now = datetime.datetime.now()
+
+    for config_name, datetime_and_snapshots in snapshots_of_configs.items():
```
However, if the return value is a tuple/list, users won't know which configuration files' snapshots will be deleted. (Optional command-line parameters such as time ranges and indicators are just ways to filter snapshots. Some snapshots belonging to configuration file X might fall within the user-selected range, while those from configuration file Y might fall outside of it.)

hirak99 commented:
I have pulled in the changes reviewed so far.

I will look forward to a PR of the [batch_deleter](https://github.com/Tql-ws1/yabsnap/commit/f0eb2a69faa94900637ccaf1a5a8465b21f020a4) commit.
